web/e2e-tests/lib/common.ts

```diff
@@ -505,16 +506,33 @@ export async function get_rendered_messages(
     return Promise.all(
         recipient_rows.map(async (element): Promise<[string, string[]]> => {
             const stream_label = await element.$(".stream_label");
-            const stream_name = (await get_element_text(stream_label!)).trim();
+            const stream_name =
+                stream_label === null ? "" : (await get_element_text(stream_label)).trim();
+            const channel_view =
+                (await page.$$(".message-header-stream-settings-button")).length > 0;
             const topic_label = await element.$(".stream_topic a");
             const topic_name =
                 topic_label === null ? "" : (await get_element_text(topic_label)).trim();
-            let key = stream_name;
-            if (topic_name !== "") {
+            let key;
+            if (channel_view && topic_name !== "") {
+                // We use only topic_name when we are in stream narrow and
+                // stream header button is visible.
+                key = topic_name;
+            } else if (topic_name !== "" && stream_name !== "") {
                 // If topic_name is '', then this is direct messages, so only
-                // append > topic_name if we are not in 1:1 or group direct
-                // messages.
+                // append > topic_name if we are in combined feed and not in
+                // 1:1 or group direct messages.
                 key = `${stream_name} > ${topic_name}`;
+            } else {
+                key = stream_name;
+            }
+            if (topic_name !== "") {
+                // If topic_name is '', then this is direct messages, so we
+                // use topic_name with possibly stream_name on combined feed.
+                key =
+                    (await page.title()) === "Combined feed - Zulip Dev - Zulip"
+                        ? `${stream_name} > ${topic_name}`
+                        : topic_name;
```

Comment:
This code is really hard to read. I suggest using fewer `?` comprehensions and using normal if/else statements, especially for situations like the topic_name being set which you check multiple times.